_posts/2015-11-25-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md

```diff
@@ -0,0 +1,197 @@
+    
+	``` fs
+        module Properties = 
+        
+            open NUnit.Framework 
+            open FsCheck
+    ```
+	
+5. Add this test to enable NCrunch
+    
+	``` fs
+            //Needed to enable NCrunch 
+            [<Ignore>][<Test>] 
+            let ignoreMe() = () 
+    ```
+	
+6. Create a static class whose methods are the properties you want to test. For this example I'll test a plainly wrong property about lists which is **the reverse of a list is always equal to the original**. 
+    
+	``` fs
+            type ListProperties =     
+                // Note: should fail     
+                static member reverseIsAsTheOriginal (xs:int[]) =          
+                    List.rev xs = xs
+    ```
+	
+7. Add the following test to verify the property defined in the ```ListProperties``` type
+    
+	``` fs
+            [<Test>] 
+            let verifyAll () =      
+                Check.QuickThrowOnFailureAll<ListProperties>()
+    ```
+	
+8. 	Enable NCrunch following *NCRUNCH > Enable NCrunch* from the main menu and let the engine catch the error
+9. If everything is in place you should see the visual error for the feedback
+	<center><img src="/assets/img/custom/blog/2015-11-25-Property-based-testing/NCrunch-visual-feed-back-error.png"></center>
+10. Run All Tests in the Resharper Runner to check the output
+	<center><img src="/assets/img/custom/blog/2015-11-25-Property-based-testing/Resharper-tests-run-error.png"></center>
+11. You should read a message similar to
+    
+	```
+        --- Checking ListProperties ---
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+		  ----> System.Exception : ListProperties.ReverseOfReverseIsAsTheOriginal-Falsifiable, after 6 tests (3 shrinks) (StdGen (1165808905,296086422)):
+		Original:
+		[-1; 2; -2]
+		Shrunk:
+		[1; 0]
+    ```
+	
+12. This shows that the property was falsifiable (as expected) but let's try to gather more informations to address the issue
+13. Change the test so that it provides a verbose output when failing
+    
+	```
+            [<Test>] 
+    		let verifyAll () = 
+                Check.VerboseThrowOnFailureAll<ListProperties>()
+    ```
+	
+14. Re-run the tests through Resharper Runner and read the output
+    
+	```
+        --- Checking ListProperties ---
+		0:
+		[-2]
+		1:
+		[-2; -2]
+		2:
+		[]
+		3:
+		[1; 1]
+		4:
+		[-2]
+		5:
+		[-3; 1; 5; -3; 6; 2; 6; -2]
+		shrink:
+		[1; 5; -3; 6; 2; 6; -2]
+		shrink:
+		[5; -3; 6; 2; 6; -2]
+		shrink:
+		[-3; 6; 2; 6; -2]
+		shrink:
+		[6; 2; 6; -2]
+		shrink:
+		[2; 6; -2]
+		shrink:
+		[6; -2]
+		shrink:
+		[6; 2]
+		shrink:
+		[6; 0]
+		shrink:
+		[3; 0]
+		shrink:
+		[2; 0]
+		shrink:
+		[1; 0]
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+		  ----> System.Exception : ListProperties.ReverseOfReverseIsAsTheOriginal-Falsifiable, after 6 tests (11 shrinks) (StdGen (1681135586,296086423)):
```

Comment:
however trivial it might sound, please explain 'Falsifiable' or put a link to it.